rust/lance-io/src/object_store/gcs_wrapper.rs

```diff
@@ -0,0 +1,458 @@
+        Ok(Self {
+            store,
+            path: Arc::new(path),
+            buffer: Vec::with_capacity(INITIAL_UPLOAD_SIZE),
+            state: UploadState::Pending,
+            connection_resets: 0,
+        })
+    }
+
+    fn put_next_part(
+        path: &Arc<Path>,
+        store: &Arc<GoogleCloudStorage>,
+        buffer: &mut Vec<u8>,
+        part_idx: &mut u16,
+        multipart_id: &Arc<MultipartId>,
+    ) -> BoxFuture<'static, std::result::Result<(u16, PartId), UploadPutError>> {
+        // Increase the upload size every 100 parts. This gives maximum part size of 2.5TB.
+        let new_capacity = ((*part_idx / 100) as usize + 1) * INITIAL_UPLOAD_SIZE;
+        let new_buffer = Vec::with_capacity(new_capacity);
+        let part = std::mem::replace(buffer, new_buffer);
+        let part = Bytes::from(part);
+
+        let part_idx_clone = *part_idx;
+        *part_idx += 1;
+        let store = store.clone();
+        let multipart_id = multipart_id.clone();
+        let path = path.clone();
+        Self::put_part(
+            path.clone(),
+            store.clone(),
+            part,
+            part_idx_clone,
+            multipart_id.clone(),
+            None,
+        )
+    }
+
+    fn put_part(
+        path: Arc<Path>,
+        store: Arc<GoogleCloudStorage>,
+        buffer: Bytes,
+        part_idx: u16,
+        multipart_id: Arc<MultipartId>,
+        sleep: Option<std::time::Duration>,
+    ) -> BoxFuture<'static, std::result::Result<(u16, PartId), UploadPutError>> {
+        Box::pin(async move {
+            if let Some(sleep) = sleep {
+                tokio::time::sleep(sleep).await;
+            }
+            let part_id = store
+                .put_part(
+                    path.as_ref(),
+                    multipart_id.as_ref(),
+                    part_idx as usize,
+                    buffer.clone(),
+                )
+                .await
+                .map_err(|source| UploadPutError {
+                    part_idx,
+                    buffer,
+                    source,
+                })?;
+            Ok((part_idx, part_id))
+        })
+    }
+
+    fn poll_tasks(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Result<(), io::Error> {
+        let mut_self = &mut *self;
+        loop {
+            match &mut mut_self.state {
+                UploadState::Pending | UploadState::Done => break,
+                UploadState::CreatingUpload(ref mut fut) => match fut.poll_unpin(cx) {
+                    Poll::Ready(Ok(multipart_id)) => {
+                        let futures = FuturesUnordered::new();
+                        let multipart_id = Arc::new(multipart_id);
+
+                        futures.push(Self::put_next_part(
+                            &mut_self.path,
+                            &mut_self.store,
+                            &mut mut_self.buffer,
```

Comment:
If `put_next_part` was a `self` method instead of a `Self` method you wouldn't need to pass all these things.